lib/collection/src/collection_manager/optimizers/config_mismatch_optimizer.rs

```diff
@@ -0,0 +1,488 @@
+    pub fn new(
+        thresholds_config: OptimizerThresholds,
+        segments_path: PathBuf,
+        collection_temp_dir: PathBuf,
+        collection_params: CollectionParams,
+        hnsw_config: HnswConfig,
+        quantization_config: Option<QuantizationConfig>,
+    ) -> Self {
+        ConfigMismatchOptimizer {
+            thresholds_config,
+            segments_path,
+            collection_temp_dir,
+            collection_params,
+            hnsw_config,
+            quantization_config,
+            telemetry_durations_aggregator: OperationDurationsAggregator::new(),
+        }
+    }
+
+    /// Calculates and HNSW config that should be used for a given vector
+    /// with current configuration.
+    ///
+    /// Takes vector-specific HNSW config (if any) and merges it with the collection-wide config.
+    fn get_required_hnsw_config(&self, vector_name: &str) -> Cow<HnswConfig> {
+        let target_hnsw_collection = &self.hnsw_config;
+        // Select vector specific target HNSW config
+        let target_hnsw_vector = self
+            .collection_params
+            .vectors
+            .get_params(vector_name)
+            .and_then(|vector_params| vector_params.hnsw_config)
+            .map(|vector_hnsw| vector_hnsw.update(target_hnsw_collection))
+            .and_then(|hnsw| match hnsw {
+                Ok(hnsw) => Some(hnsw),
+                Err(err) => {
+                    log::warn!(
+                        "Failed to merge collection and vector HNSW config, ignoring: {err}"
+                    );
+                    None
+                }
+            });
+        match target_hnsw_vector {
+            Some(target_hnsw) => Cow::Owned(target_hnsw),
+            None => Cow::Borrowed(target_hnsw_collection),
+        }
+    }
+
+    fn worst_segment(
+        &self,
+        segments: LockedSegmentHolder,
+        excluded_ids: &HashSet<SegmentId>,
+    ) -> Vec<SegmentId> {
+        let segments_read_guard = segments.read();
+        let candidates: Vec<_> = segments_read_guard
+            .iter()
+            // Excluded externally, might already be scheduled for optimization
+            .filter(|(idx, _)| !excluded_ids.contains(idx))
+            .filter_map(|(idx, segment)| {
+                let segment_entry = segment.get();
+                let read_segment = segment_entry.read();
+                let point_count = read_segment.available_point_count();
+                let vector_size = point_count
+                    * read_segment
+                        .vector_dims()
+                        .values()
+                        .max()
+                        .copied()
+                        .unwrap_or(0)
+                    * VECTOR_ELEMENT_SIZE;
+
+                let segment_config = read_segment.config();
+
+                if read_segment.segment_type() == SegmentType::Special {
+                    return None; // Never optimize already optimized segment
```

Comment:
Isn't this comment referring to `SegmentType::Indexed`?

Reply:
it is referring to the segment which is currently already in the process of optimization